### ISSUE TITLE
Menu Widget ItemHeight/Width Dynamic

### ISSUE THREAD

**Issue by genusistimelord** [iced-rs/iced_aw#158]
Would be nice to Add a dynamic ItemHeight and or Width. This would at least allow the menu to adapt to the controls rather than the controls adapting to the menu. easier to setup with less issue of Clipping text etc. @latidoremi

**Comment by latidoremi** [comment]
Dynamic item height is doable, but I'm not so sure with dynamic item width

For dynamic width to work, the minimum width should be available from every item in order to compute a proper width of a menu. `Widget::width` is not useful in this case, if all of the items in a menu have `Length::Fill` widths, we can only fallback to a default width which is not that different than setting a width manually. Even if some of the items have `Length::Fixed` widths, with which we can compute a proper width for them and have other items comply to it, there'll be gaps when their widths are different

<img width="126" alt="gaps" src="https://github.com/iced-rs/iced_aw/assets/49757619/00f74e8d-2e60-42fe-9f79-e7c09f79f1d9">

So unless we have something like `Widget::get_min_size`, I don't think dynamic item width is possible in the current design

Alternatively we can go with Kaiden42's design, where only certain types of item are allowed, practically it's just converting button, checkbox, text, etc. into the menu item version, and we can define a min width for each of them, but that'll be a different menu widget

**Comment by genusistimelord** [comment]
I think with dynamic width the best thing to do is maybe have a maximum size width limit. Also the reason I am suggesting these is due to the nature of how fonts and such work now. Fonts no longer use the Width and height of the Box they are in instead its split up into  Font size and line_height. which messes up the rendering if you set the Height to be statically smaller than the font size *  line height. so it causes clipping issues. I did fix the clipping issues in the example by making the size Bigger, but it would be nice to have a Max and Min limit and then allow Dynamic, static and uniform to be based around these.

**Comment by latidoremi** [comment]
So basically
```
let item_width = match widget_bounds_width{
    Length::Fill => static_width,
    Length::Fixed(x) => x.max(static_width)
};
let item_height = match widget_bounds_height{
    Length::Fill => static_height,
    Length::Fixed(x) => x.max(static_height)
};
```
Is that correct?

But what does max size limit has to do with the clipping issues?

**Comment by genusistimelord** [comment]
that would work but I think the issue with Fill is the Max would need to Clip the Fill so it cant go past a certain Point. Especially for width. Height I'm not to concerned here about but it would be nice to have some overall limitation even if its huge by default.

the Max size limit is only to Prevent the width from growing to large and taking up the entire screen. so it would help limit the overall size anything could become. you could make the max size a % of the screen-size by default. or make Max its own Type that has a % base or a Static set base. but the issue with Font clipping is If you make the Box to small it will clip the font now.

So the issue with Fonts shows up like this if the height is set to small. Dynamic height would fix this. width is not as important here since the height would just adapt and newline the text.
![image](https://github.com/iced-rs/iced_aw/assets/1034709/7ebd419a-717c-4717-84a9-9b5a3cdf1314)

now if i change the size to something like 30 it then is fine. this is the new Font system using Cosmic-text in Iced now. if your wondering what the major change is. 

So to sum it up

Max size's would be to limit the menus from overgrowing in dynamic mode.
Min Size's would be to ensure uniform or dynamic cant go below its limit. this could be based on Font size * line height.
Adding a Dynamic Fill type to auto adjust to actual control sizes. limited by the above.

**Comment by latidoremi** [comment]
I still don't see the point of adding a `Dynamic` variant to `ItemWidth`, the real dynamic width is not achievable rn, Static/Uniform already clamp Fill width with the value set, so we're adding a % based value to clamp Static/Uniform? How is that useful in any case? Resizing the window?

I do understand the use case of `ItemHeight::Dynamic`, I can work on that, but how do I get font size*line height in `fn layout`? Also this limit should be optional since not every widget has text

**Comment by genusistimelord** [comment]
if you can make sure the padding doesn't decrease the return size of the elements layout , that probably would fix the issue of clipping. and instead make the storage area larger based on that layout size and the padding for dynamic. Also you don't have to do width. It was more of What if i wanted to have a dynamic scalable width up to a Set size. where my Menu width is generally Length::Shrink but the element i want to add in is much larger with a length::Fill. to just Expand it to that. just for menu it doesn't make sense in most cases to have something that big. 

for font stuff layouts should auto calculate them for their returns. You just need to handle how padding and such is done differently per element. If you do need to get font stuff you can only get it for your current widget thats handling the layout. Please see other widgets that do so in iced_aw.
